src/app/clusters/mode-base-server/mode-base-server.cpp

```diff
@@ -0,0 +1,445 @@
+    if (!emberAfIsKnownVolatileAttribute(mEndpointId, mClusterId, Attributes::CurrentMode::Id) &&
+        !emberAfIsKnownVolatileAttribute(mEndpointId, mClusterId, Attributes::StartUpMode::Id))
+    {
+        // Read the StartUpMode attribute and set the CurrentMode attribute
+        // The StartUpMode attribute SHALL define the desired startup behavior of a
+        // device when it is supplied with power and this state SHALL be
+        // reflected in the CurrentMode attribute.  The values of the StartUpMode
+        // attribute are listed below.
+
+        DataModel::Nullable<uint8_t> startUpMode = GetStartUpMode();
+        if (!startUpMode.IsNull())
+        {
+#ifdef EMBER_AF_PLUGIN_ON_OFF
+            // OnMode with Power Up
+            // If the On/Off feature is supported and the On/Off cluster attribute StartUpOnOff is present, with a
+            // value of On (turn on at power up), then the CurrentMode attribute SHALL be set to the OnMode attribute
+            // value when the server is supplied with power, except if the OnMode attribute is null.
+            if (emberAfContainsServer(mEndpointId, OnOff::Id) &&
+                emberAfContainsAttribute(mEndpointId, OnOff::Id, OnOff::Attributes::StartUpOnOff::Id) &&
+                emberAfContainsAttribute(mEndpointId, mClusterId, ModeBase::Attributes::OnMode::Id) &&
+                HasFeature(ModeBase::Feature::kOnOff))
+            {
+                DataModel::Nullable<uint8_t> onMode = GetOnMode();
+                bool onOffValueForStartUp           = false;
+                if (!emberAfIsKnownVolatileAttribute(mEndpointId, OnOff::Id, OnOff::Attributes::StartUpOnOff::Id) &&
+                    OnOffServer::Instance().getOnOffValueForStartUp(mEndpointId, onOffValueForStartUp) == EMBER_ZCL_STATUS_SUCCESS)
+                {
+                    if (onOffValueForStartUp && !onMode.IsNull())
+                    {
+                        ChipLogDetail(Zcl, "ModeBase: CurrentMode is overwritten by OnMode");
+                        // it is overwritten by the on/off cluster.
+                        return CHIP_NO_ERROR;
+                    }
+                }
+            }
+#endif // EMBER_AF_PLUGIN_ON_OFF
+
+            BootReasonType bootReason = BootReasonType::kUnspecified;
+            CHIP_ERROR error          = DeviceLayer::GetDiagnosticDataProvider().GetBootReason(bootReason);
+
+            if (error != CHIP_NO_ERROR)
+            {
+                ChipLogError(Zcl, "Unable to retrieve boot reason: %" CHIP_ERROR_FORMAT, error.Format());
+                // We really only care whether the boot reason is OTA.  Assume it's not.
+                bootReason = BootReasonType::kUnspecified;
+            }
+            if (bootReason == BootReasonType::kSoftwareUpdateCompleted)
+            {
+                ChipLogDetail(Zcl, "ModeBase: CurrentMode is ignored for OTA reboot");
+                return CHIP_NO_ERROR;
+            }
```

Comment:
This only works if the device is using Matter OTA which **not all devices do**. It also only works if the bootreason for OTA is managed properly in Diagnostics Data Provider. This seems risky and could lead to a false sense of security if it is required to support to deal with version downgrades.

Reply:
Is there a suggestion for what to do here? The spec asks for this behaviour. Is there a better way to ensure that we pickup the reason for boot?

Reply:
OK. I will note this down. I suggest that this is also noted in the MadeSelect cluster as the spec and SDK implementation here are the same.

Reply:
`todo this only works for matter OTAs. According to the spec, this should also work for general OTAs.`